example/src/main/java/io/netty/example/dns/udp/DnsClient.java

```diff
@@ -0,0 +1,106 @@
+import io.netty.handler.codec.dns.DnsQuery;
+import io.netty.handler.codec.dns.DnsQuestion;
+import io.netty.handler.codec.dns.DnsRawRecord;
+import io.netty.handler.codec.dns.DnsRecord;
+import io.netty.handler.codec.dns.DnsRecordType;
+import io.netty.handler.codec.dns.DnsSection;
+import io.netty.util.NetUtil;
+import io.netty.util.internal.SocketUtils;
+
+public class DnsClient {
+
+    private static final String QUERY_DOMAIN = "www.example.com";
+    private static final int DNS_SERVER_PORT = 53;
+    private static final String DNS_SERVER_HOST = "8.8.8.8";
+
+    private static void handleQueryResp(DatagramDnsResponse msg) {
+        if (msg.count(DnsSection.QUESTION) > 0) {
+            DnsQuestion question = msg.recordAt(DnsSection.QUESTION, 0);
+            System.out.printf("name: %s%n", question.name());
+        }
+        for (int i = 0, count = msg.count(DnsSection.ANSWER); i < count; i++) {
+            DnsRecord record = msg.recordAt(DnsSection.ANSWER, i);
+            if (record.type() == DnsRecordType.A) {
+                //just print the IP after query
+                DnsRawRecord raw = (DnsRawRecord) record;
+                System.out.println(NetUtil.bytesToIpAddress(ByteBufUtil.getBytes(raw.content())));
+            }
+        }
+    }
+
+    public static void main(String[] args) throws Exception {
+        InetSocketAddress addr = SocketUtils.socketAddress(DNS_SERVER_HOST, DNS_SERVER_PORT);
+        final CountDownLatch latch = new CountDownLatch(1);
+        EventLoopGroup group = new NioEventLoopGroup();
+        try {
+            Bootstrap b = new Bootstrap();
+            b.group(group)
+             .channel(NioDatagramChannel.class)
+             .handler(new ChannelInitializer<DatagramChannel>() {
+                 @Override
+                 protected void initChannel(DatagramChannel ch) throws Exception {
+                     ChannelPipeline p = ch.pipeline();
+                     p.addLast(new DatagramDnsQueryEncoder())
+                     .addLast(new DatagramDnsResponseDecoder())
+                     .addLast(new SimpleChannelInboundHandler<DatagramDnsResponse>() {
+                        @Override
+                        protected void channelRead0(ChannelHandlerContext ctx, DatagramDnsResponse msg)
+                                throws Exception {
+                            try {
+                                handleQueryResp(msg);
+                            } finally {
+                                latch.countDown();
+                            }
+                        }
+                    });
+                 }
+             });
+            Channel ch = b.bind(0).sync().channel();
+            DnsQuery query = new DatagramDnsQuery(null, addr, 1).setRecord(
+                    DnsSection.QUESTION,
+                    new DefaultDnsQuestion(QUERY_DOMAIN, DnsRecordType.A));
+            ch.writeAndFlush(query);
+            latch.await(10L, TimeUnit.SECONDS);
```

Comment:
you could also just call `ch.closeFuture().sync()` and call `ch.close()` in your `channelRead0` method